README.md
# episode-parser

[![CircleCI](https://img.shields.io/circleci/project/tregusti/episode-parser/master.svg)][circleci]
[![NPM release](https://img.shields.io/npm/v/episode-parser.svg)][npm-module]
[![License](https://img.shields.io/npm/l/episode-parser.svg)][license]

  [circleci]: https://circleci.com/gh/tregusti/episode-parser
  [npm-module]: https://www.npmjs.com/package/episode-parser
  [license]: http://tregusti.mit-license.org/

A javascript utility for parsing file names in a format that sometimes is used
for tv shows.

It parses a file name like this `Fake.Blood.2013.S07E06.Karma.Sucks.PROPER.720p.HDTV.x264-STALKERS.srt`
into an object with the following possible properties:

  - `show`, string: The name of the show, `'Fake Blood'`.
  - `year`, integer: The release year, `2013`.
  - `season`, integer: The season number, `7`.
  - `episode`, integer: The episode number, `6`.
  - `episodeCount`, integer: The count of episodes this file name includes, `1`.
  - `name`, string: The name of the episode, `'Karma Sucks'`.
  - `quality`, integer: The quality taken from `720p` or `1080p`. In this case `720`.
  - `source`, string: Available values are `'hdtv'`, `'webdl'`, In this case `'hdtv'`.
  - `codec`, string: Codec used, `'x264'`.
  - `group`, string: Release group, `'STALKERS'`.
  - `ext`, string: File extension, `'srt'`.

## Supported formats

Take a look in [the tests][examples] for examples of what file name formats are
supported.

  [examples]: https://github.com/tregusti/episode-parser/blob/master/test/parser.test.js

## Usage

Simply require the module and you get a parsing function that you invoke with
the file name to parse:

```js
var parser = require('episode-parser')
var filename = 'Fake Blood 2x06 I Need Blood.srt'
var result = parser(filename)
console.log(result.show) //= 'Fake Blood'
console.log(result.year) //= undefined
console.log(result.season) //= 2
console.log(result.codec) //= undefined
```

If the parser can't parse the file name, `null` will be returned. If the parser
can't parse out a specific details, the property for that detail will be missing
from the result.

## Installation

```sh
npm install episode-parser
```

## Versioning

This package uses [semver](http://semver.org/spec/v2.0.0.html).

## Contributing

If you feel something is missing, which is highly probable since I made this for
my personal needs, please create a [pull request][pr-help], with tests covering
your case. If you feel somewhat lazy, create an issue instead and I might look
into it.

  [pr-help]: https://help.github.com/articles/using-pull-requests/

## Thank you

[![Flattr](http://img.shields.io/badge/flattr-donate-brightgreen.svg)](https://flattr.com/profile/tregusti)

Building good software and tools takes time. Please do support a future for this project if you are
using it. Thank you!


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.0.2](https://github.com/tregusti/episode-parser/compare/v2.0.0...v2.0.2) (2021-04-22)


### Bug Fixes

* Bump deps ([4c2b226](https://github.com/tregusti/episode-parser/commit/4c2b226bb430a95326b61d20397155fec038ae43))

### [2.0.0](https://github.com/tregusti/episode-parser/compare/v2.0.0...v1.5.0) (2020-07-15)

* Removed support of Node v8. Rerelease of 1.5.1 since it was a breaking change.

### [1.5.1](https://github.com/tregusti/episode-parser/compare/v1.5.0...v1.5.1) (2020-07-15)

* **Bad release** Bump deps to latest version

## [1.5.0](https://github.com/tregusti/episode-parser/compare/v1.4.0...v1.5.0) (2020-01-04)


### Features

* Expose DVD as source in result ([17dbe93](https://github.com/tregusti/episode-parser/commit/17dbe93)), closes [#13](https://github.com/tregusti/episode-parser/issues/13)
* Parse 265 codec ([a6a8f55](https://github.com/tregusti/episode-parser/commit/a6a8f55)), closes [#12](https://github.com/tregusti/episode-parser/issues/12)



## [1.4.0](https://github.com/tregusti/episode-parser/compare/v1.3.0...v1.4.0) (2019-07-15)


### Features

* Add typescript support ([067db26](https://github.com/tregusti/episode-parser/commit/067db26))



## [1.3.0](https://github.com/tregusti/episode-parser/compare/v1.2.0...v1.3.0) (2019-07-15)

### Features

- Double digit seasons in previous 000 format ([8ea338e](https://github.com/tregusti/episode-parser/commit/8ea338e)), closes [#6](https://github.com/tregusti/episode-parser/issues/6)
- Add `episodeCount` to result ([1e9658c](https://github.com/tregusti/episode-parser/commit/1e9658c))
- Parse english like file names ([51dad93](https://github.com/tregusti/episode-parser/commit/51dad93)) @Jephuff


test/parser.test.js
'use strict'

var example = require('./support/example')

describe('parser', function() {
  example('Hello 2014 Season 01 Episode 1 the Problem 720p.HDTV.x264-SKGTV.srt', {
    show:    'Hello',
    season:  1,
    episode: 1,
    name: 'the Problem',
    year: 2014,
    quality: 720,
    source: 'hdtv',
    codec: 'x264',
    group: 'SKGTV',
    ext: 'srt'
  })
  example('Hello 2014 Season 01 Episode 1-4 the Problem 720p.HDTV.x264-SKGTV.srt', {
    show:    'Hello',
    season:  1,
    episode: 1,
    name: 'the Problem',
    year: 2014,
    quality: 720,
    source: 'hdtv',
    codec: 'x264',
    group: 'SKGTV',
    ext: 'srt',
    episodeCount: 4
  })
  
  example('comedians in cars getting coffee S1 E1', {
    show:    'comedians in cars getting coffee',
    season:  1,
    episode: 1,
  })
  example('battlestar galactica s01e01-e03 - saga of a star world.mp4', {
    show:    'battlestar galactica',
    season:  1,
    episode: 1,
    episodeCount: 3,
  })

  example('full frontal with samantha bee s01e14.mkv', {
    show:    'full frontal with samantha bee',
    season:  1,
    episode: 14,
    episodeCount: 1,
  })

  example('Goodbye.S01E08.DVDRip-S0ME', {
    show:    'Goodbye',
    season:  1,
    episode: 8,
    source:  'dvd',
  })
  example('Hello.S01E05.HDTV-GOTTALOVEiT', {
    show:    'Hello',
    season:  1,
    episode: 5,
    source:  'hdtv',
    group:   'GOTTALOVEiT'
  })
  example('Young.Master.S02E01.720p.HDTV-GOTTALOVEiT', {
    show:    'Young Master',
    season:  2,
    episode: 1,
    quality: 720,
    source:  'hdtv',
    group:   'GOTTALOVEiT'
  })
  example('Da.Vincis.Demons.S02E05.HDTV.x264-KILLERS.mp4', {
    show:    'Da Vincis Demons',
    season:  2,
    episode: 5,
    source:  'hdtv',
    codec:   'x264',
    group:   'KILLERS',
    ext:     'mp4'
  })
  example('House.of.Cards.2013.S02E06.720p.HDTV.x264-SKGTV.srt', {
    show:    'House of Cards',
    season:  2,
    episode: 6,
    year:    2013,
    quality: 720,
    source:  'hdtv',
    codec:   'x264',
    group:   'SKGTV',
    ext:     'srt'
  })
  example('girls.s01e04.720p.hdtv.x264-2hd.mkv', {
    show: 'girls',
    season: 1,
    episode: 4
  })
  example('Orange Is the New Black S01E01 I Wasn\'t Ready 720p BRRip DD5.1 x264-PSYPHER.mkv', {
    show: 'Orange Is the New Black',
    season: 1,
    episode: 1,
    name: 'I Wasn\'t Ready',
    quality: 720,
    source: 'bluray',
    codec: 'x264',
    group: 'PSYPHER',
    ext: 'mkv'
  })
  example('Monster.1x19.720p.HDTV.x264-HiPP', {
    show: 'Monster',
    season: 1,
    episode: 19,
    source: 'hdtv',
    codec: 'x264',
    group: 'HiPP'
  })

  example('Monster.S01E01.264.days.of.horror.720p.mkv', {
    source: undefined,
    codec: undefined,
    name: '264 days of horror'
  })
  describe('codec', function() {
    example('Show.S01E01.Name.720p.H.264-Group', {
      show: 'Show',
      name: 'Name',
      codec: 'x264',
    })
    example('Show.S01E01.Name.720p.H.265-Group', {
      show: 'Show',
      name: 'Name',
      codec: 'x265',
    })
    example('Show.S01E01.Name.720p.x264-Group', {
      show: 'Show',
      name: 'Name',
      codec: 'x264',
    })
    example('Show.S01E01.Name.720p.x265-Group', {
      show: 'Show',
      name: 'Name',
      codec: 'x265',
    })
  })
  example('Hannibal.S02E09.Shizakana.720p.WEB-DL.DD5.1.H.264-NTb', {
    show: 'Hannibal',
    name: 'Shizakana',
    source: 'webdl',
    codec: 'x264',
    group: 'NTb'
  })
  example('True.Blood.S07E06.Karma.PROPER.720p.HDTV.x264-KILLERS', {
    show: 'True Blood',
    season: 7,
    episode: 6,
    name: 'Karma',
    quality: 720,
    source: 'hdtv',
    codec: 'x264',
    group: 'KILLERS'
  })
  example('Fairly Legal S02E02 Start Me Up.mkv', {
    show: 'Fairly Legal',
    season: 2,
    episode: 2,
    name: 'Start Me Up',
    ext: 'mkv'
  })
  example('Orphan.Black.S03E05.Scarred.by.Many.Past.Frustrations.720p.HDTV.x264-KILLERS[rarbg]', {
    group: 'KILLERS',
    name: 'Scarred by Many Past Frustrations'
  })
  example('Futurama.S06.1080p.BluRay.DTS.x264-CtrlHD', {
    show: 'Futurama',
    season: 6,
    quality: 1080,
    source: 'bluray',
    codec: 'x264',
    group: 'CtrlHD'
  })
  example('the.good.wife.709.hdtv-lol', {
    show: 'The Good Wife',
    season: 7,
    episode: 9,
    source: 'hdtv',
    group: 'lol'
  })
  example('ray.donovan.312.720p-dimension', {
    show: 'Ray Donovan',
    season: 3,
    episode: 12,
    quality: 720,
    group: 'dimension'
  })
  example('the.big.bang.theory.1017.hdtv-lol', {
    show: 'The Big Bang Theory',
    season: 10,
    episode: 17,
    source: 'hdtv',
    group: 'lol'
  })
  example('the.big.bang.theory.1017-19.hdtv-lol', {
    show: 'The Big Bang Theory',
    season: 10,
    episode: 17,
    episodeCount: 3,
    source: 'hdtv',
    group: 'lol'
  })
  example('Homeland - 03x09 - Horse and Wagon', {
    show: 'Homeland',
    season: 3,
    episode: 9,
    name: 'Horse and Wagon'
  })
  example('Homeland - 03x09-10 - Horse and Wagon', {
    show: 'Homeland',
    season: 3,
    episode: 9,
    episodeCount: 2,
    name: 'Horse and Wagon'
  })
})


test/support/example.js
'use strict'

var expect = require('chai').expect

function run(filename, details, context) {
  context('with filename like ' + filename, function() {
    var result
    beforeEach(function() {
      result = require('../..')(filename)
    })
    Object.keys(details).forEach(function(property) {
      var value = details[property]
      /* eslint curly:0 */
      if (value === undefined) {
        it('does not find the ' + property, function() {
          expect(result).to.not.have.a.property(property)
        })
      } else {
        it('finds the ' + property, function() {
          expect(result).to.have.a.property(property, value)
        })
      }
    })
  })
}

module.exports = function(filename, details) {
  run(filename, details, context)
}
module.exports.only = function(filename, details) {
  run(filename, details, context.only)
}
module.exports.skip = function(filename, details) {
  run(filename, details, context.skip)
}


lib/parser.js
'use strict'

module.exports = function(filename) {
  var result

  if (result = parseEnglishLike(filename)) return result
  if (result = parseSceneLike(filename)) return result
  if (result = parseSceneLikeWithX(filename)) return result
  if (result = parseSceneLike000(filename)) return result

  return null
}

function parseEnglishLike(filename) {
  var m, result = { episodeCount: 1 }

  if (m = filename.match(/^(.+?)([\. ])(?:((?:20|19)\d\d)\2)?Season (\d+) (Episode (\d+))?(-(\d+))?/i)) {
    var show = m[1].replace(/\./g, ' ')
    result.show =    show
    result.season =  +m[4]
    if (m[6]) result.episode = +m[6]

    if (m[3]) result.year = +m[3]

    if (m[8]) result.episodeCount += +m[8] - result.episode

    parseDetails(filename.substr(m[0].length), result)

    return result
  }
}

function parseSceneLike(filename) {
  var m, result = { episodeCount: 1 }

  if (m = filename.match(/^(.+?)([\. ])(?:((?:20|19)\d\d)\2)?S(\d+) ?(E(\d+))?(-(E(\d+)))?/i)) {
    var show = m[1].replace(/\./g, ' ')
    result.show =    show
    result.season =  +m[4]
    if (m[6]) result.episode = +m[6]

    if (m[3]) result.year = +m[3]

    if (m[9]) result.episodeCount += +m[9] - result.episode

    parseDetails(filename.substr(m[0].length), result)

    return result
  }
}

function parseSceneLikeWithX(filename) {
  var m, result = { episodeCount: 1 }

  if (m = filename.match(/^(.+?)([\. ])(?:((?:20|19)\d\d)\2)?(\d+)x(\d+)(-(\d+))?\2/i)) {
    var show = humanize(m[1])
    result.show =    show
    result.season =  +m[4]
    result.episode = +m[5]

    if (m[7]) result.episodeCount += +m[7] - result.episode

    if (m[3]) result.year = +m[3]

    parseDetails(filename.substr(m[0].length), result)

    return result
  }
}
function parseSceneLike000(filename) {
  var m, result = { episodeCount: 1 }

  if (m = filename.match(/^(.+?)([\. ])(?:((?:20|19)\d\d)\2)?(\d\d\d+)(-(\d+))?\2/i)) {
    var show = humanize(m[1])
      .replace(/\w\S*/g, function(txt){return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase()})
    result.show =    show
    var episode = m[4].substr(-2)
    result.season =  +m[4].substring(0, m[4].length - episode.length)
    result.episode = +episode

    if (m[6]) result.episodeCount += +m[6] - result.episode

    if (m[3]) result.year = +m[3]

    parseDetails(filename.substr(m[0].length), result)

    return result
  }
}
function humanize(str) {
  return str.replace(/\./g, ' ').replace(/(^- )|( -$)/, '')
}

function parseDetails(details, result) {
  var re = {
    quality:   /(480|720|1080)[pi]/,
    source:    /(hdtv|brrip|bluray|bd|dvd|dvdrip|web-?dl)/i,
    codec:     /((?:h\.|x)264|x265)/i,
    other:     /((?:real[\. ])?proper)/ig,
    group:     /-(\w+)(?:\[\w+\])?(\....)?$/,
    extension: /\.(...)$/
  }
  re.all = new RegExp(Object.keys(re).map(function(name) {
    return re[name].source
  }).join('|'), 'i')

  var reStripper = new RegExp('(.*?)(' + re.all.source + ').*$', 'i')
  var name = humanize(details.replace(reStripper, '$1')).trim()
  if (name) result.name = name

  var quality = details.match(re.quality)
  if (quality) result.quality = +quality[1]

  var source = details.match(re.source)
  if (source) {
    source = source[0].toLowerCase()
    if (source.substr(0, 1) === 'b') result.source = 'bluray'
    if (source === 'hdtv') result.source = 'hdtv'
    if (/web-?dl/.test(source)) result.source = 'webdl'
  }

  var codec = details.match(re.codec)
  if (codec) {
    codec = codec[0].toLowerCase()
    if (codec.indexOf('264') >= 0) codec = 'x264'
    result.codec = codec
  }

  var group = details.match(re.group)
  if (group) result.group = group[1]

  var extension = details.match(re.extension)
  if (extension) result.ext = extension[1]
}


lib/parser.ts
enum Source {
  BluRay = 'bluray',
  HDTV = 'hdtv',
  WebDL = 'webdl',
  DVD = 'dvd'
}

type Result = {
  show: string;
  season: number;
  year?: number;
  name?: string;
  codec?: string;
  quality?: number;
  source?: Source;
  group?: string;
  ext?: string;
} & (
  | {
    episode: number;
    episodeCount: number;
  }
  | {
    episode?: undefined;
    episodeCount?: undefined;
  })

const parsers = [
  parseEnglishLike,
  parseSceneLike,
  parseSceneLikeWithX,
  parseSceneLike000,
]
export = function(filename: string) {
  for (let i = 0; i < parsers.length; i += 1) {
    const result = parsers[i](filename)
    if (result) return result
  }
  return null
}

function parseEnglishLike(filename: string) {
  const m = filename.match(
    /^(.+?)([\. ])(?:((?:20|19)\d\d)\2)?Season (\d+) (Episode (\d+))?(-(\d+))?/i
  )

  if (m) {
    const firstEpisode = +m[6]
    const lastEpisode = m[8] ? +m[8] : firstEpisode
    const result: Result = {
      show: m[1].replace(/\./g, ' '),
      season: +m[4],
      ...(m[6]
        ? {
          episode: firstEpisode,
          episodeCount: lastEpisode - firstEpisode + 1,
        }
        : {}),
    }

    if (m[3]) result.year = +m[3]

    parseDetails(filename.substr(m[0].length), result)

    return result
  }
}

function parseSceneLike(filename: string) {
  const m = filename.match(
    /^(.+?)([\. ])(?:((?:20|19)\d\d)\2)?S(\d+) ?(E(\d+))?(-(E(\d+)))?/i
  )

  if (m) {
    const firstEpisode = +m[6]
    const lastEpisode = m[9] ? +m[9] : firstEpisode
    const result: Result = {
      show: m[1].replace(/\./g, ' '),
      season: +m[4],
      ...(m[6]
        ? {
          episode: firstEpisode,
          episodeCount: lastEpisode - firstEpisode + 1,
        }
        : {}),
    }

    if (m[3]) result.year = +m[3]

    parseDetails(filename.substr(m[0].length), result)

    return result
  }
}

function parseSceneLikeWithX(filename: string) {
  const m = filename.match(
    /^(.+?)([\. ])(?:((?:20|19)\d\d)\2)?(\d+)x(\d+)(-(\d+))?\2/i
  )

  if (m) {
    const firstEpisode = +m[5]
    const lastEpisode = m[7] ? +m[7] : firstEpisode
    const result: Result = {
      show: humanize(m[1]),
      season: +m[4],
      episode: firstEpisode,
      episodeCount: lastEpisode - firstEpisode + 1,
    }

    if (m[3]) result.year = +m[3]

    parseDetails(filename.substr(m[0].length), result)

    return result
  }
}

function parseSceneLike000(filename: string) {
  const m = filename.match(
    /^(.+?)([\. ])(?:((?:20|19)\d\d)\2)?(\d\d\d+)(-(\d+))?\2/i
  )

  if (m) {
    const episodeStr = m[4].substr(-2)
    const firstEpisode = +episodeStr
    const lastEpisode = m[6] ? +m[6] : firstEpisode
    const result: Result = {
      show: humanize(m[1]).replace(
        /\w\S*/g,
        txt => txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase()
      ),
      season: +m[4].substring(0, m[4].length - episodeStr.length),
      episode: firstEpisode,
      episodeCount: lastEpisode - firstEpisode + 1,
    }

    if (m[3]) result.year = +m[3]

    parseDetails(filename.substr(m[0].length), result)

    return result
  }
}

function humanize(str: string) {
  return str.replace(/\./g, ' ').replace(/(^- )|( -$)/, '')
}

function parseDetails(details: string, result: Result) {
  const re = {
    quality: /(480|720|1080)[pi]/,
    source: /(hdtv|brrip|bluray|bd|dvd|dvdrip|web-?dl)/i,
    codec: /((?:h\.|x)264|(?:h\.|x)265)/i,
    other: /((?:real[\. ])?proper)/gi,
    group: /-(\w+)(?:\[\w+\])?(\....)?$/,
    extension: /\.(...)$/,
  }

  const reAll = new RegExp(
    (Object.keys(re) as Array<keyof typeof re>)
      .map(name => re[name].source)
      .join('|'),
    'i'
  )

  const reStripper = new RegExp('(.*?)(' + reAll.source + ').*$', 'i')
  const name = humanize(details.replace(reStripper, '$1')).trim()
  if (name) result.name = name

  const quality = details.match(re.quality)
  if (quality) result.quality = +quality[1]

  const sourceMatch = details.match(re.source)
  if (sourceMatch) {
    const source = sourceMatch[0].toLowerCase()
    if (source.substr(0, 1) === 'b') result.source = Source.BluRay
    if (source === 'hdtv') result.source = Source.HDTV
    if (/web-?dl/.test(source)) result.source = Source.WebDL
    if (/dvd/i.test(source)) result.source = Source.DVD
  }

  const codecMatch = details.match(re.codec)
  if (codecMatch) {
    const codec = codecMatch[0].toLowerCase()
    if (codec.indexOf('264') >= 0) result.codec = 'x264'
    else if (codec.indexOf('265') >= 0) result.codec = 'x265'
    else result.codec = codec
  }

  const group = details.match(re.group)
  if (group) result.group = group[1]

  const extension = details.match(re.extension)
  if (extension) result.ext = extension[1]
}
